06x-new-kafka-roller.md

```diff
@@ -0,0 +1,350 @@
+
+### Node State
+When a new reconciliation starts up, a context object is created for each node to store the state and other useful information used by the roller. It will have the following fields:
+
+ - <i>nodeRef</i>: NodeRef object that contains Node ID.
+ - <i>currentNodeRole</i>: Currently assigned process roles for this node (e.g. controller, broker).
+ - <i>lastKnownState</i>: It contains the last known state of the node based on information collected from the abstracted sources (Kubernetes API, KafkaAgent and Kafka Admin API). The table below describes the possible states.
+ - <i>restartReason</i>: It is updated based on the current predicate logic from the `Reconciler`. For example, an update in the Kafka CR is detected.
+ - <i>numRestartAttempts</i>: The value is incremented each time the node has been attempted to restart.
+ - <i>numReconfigAttempts</i>: The value is incremented each time the node has been attempted to reconfigure.
+ - <i>numRetries</i>: The value is incremented each time the node cannot be restarted/reconfigured due to not meeting safety conditions (more on this later).
+ - <i>lastTransitionTime</i>: System.nanoTime of last observed state transition.
+
+ The following table illustrates the proposed node states and the possible transitions:
+ | State            | Description      | Possible transitions |
+ | :--------------- | :--------------- | :----------- |
+ | UNKNOWN               | The initial state when creating `Context` for a node. We expect to transition from this state fairly quickly after creating the context for nodes.  | `NOT_RUNNING` `NOT_READY` `RECOVERING` `SERVING` |
+ | NOT_RUNNING           | Node is not running (Kafka process is not running). This is determined via Kubernetes API, more details for it below. | `NOT_READY`, `NOT_RUNNING`. | `RESTARTED` `SERVING` | 
+ | NOT_READY             | Node is running but not ready to serve requests which is determined by Kubernetes readiness probe (broker state < 2 OR == 127 OR controller is not listening on port). | `RESTARTED` `SERVING` |
+ | RESTARTED             | After successful `kubectl delete pod`. | `NOT_RUNNING` `NOT_READY` `RECOVERING` `SERVING` |
+ | RECONFIGURED          | After successful Kafka node config update via Admin client. | `NOT_RUNNING` `NOT_READY` `RESTARTED` `SERVING` |
+ | RECOVERING            | Node has started but is in log recovery (broker state == 2). | `SERVING` |
+ | SERVING               | Node is in running state and ready to serve requests which is determined by Kubernetes readiness probe (broker state >= 3 AND != 127 OR controller is listening on port). | `RESTARTED` `RECONFIGURED` `LEADING_ALL_PREFERRED` |
+ | LEADING_ALL_PREFERRED | Node is in running state and leading all preferred replicas. | None
```

Comment:
Also how does it map with SERVING (or READY) state? I would expect the node to be ready and serving. The fact it's "leading all preferred" is not a node state imho, more part of the node context.

Reply:
> No transition states?

LEADING_ALL_PREFERRED is the final desired state, once the pod reaches this state, the roller won't do anything with it and complete the reconciliation if all pods are in this state. However, it is possible for the pod to become unhealthy and go into `NOT_READY`, `NOT_RUNNING` states too. So I added those as possible states transition into. 

> Also, what exactly does this mean?
That this node is leader only for the partitions where it is the preferred leader?
That all partitions where this node is the preferred leader have a leader in this node?
In other words, a ndoe that is leader for a partition for which it is not the preferred leader - is it LEADING_ALL_PREFERRED? I think this has an impact on the amount of disruption caused down the road by the roller and should be clarified.

This state is observed for pods individually. I believe it means that the particular pod is leading all the partitions which it is the preferred leader for, but it could still be leading other partitions that it is not the preferred leader for too. After restarting a node, we would wait for it to lead the partitions that they are preferred leader for. In the end, the desire is that all nodes have `LEADING_ALL_PREFERRED` state, which means  that all partitions are being led by the preferred leaders. 

> Also how does it map with SERVING (or READY) state? I would expect the node to be ready and serving. The fact it's "leading all preferred" is not a node state imho, more part of the node context.

Yes, a node can transition into `LEADING_ALL_PREFERRED`, only if it is already in `READY` state  as described in the "Possible transitions" column. It is a state in KafkaRoller sense similar to `RESTARTED` and `RECONFIGURED`, but not a state that matches the system such as Kubernetes states or Broker States. I think we need to discuss and agree what we want to consider as states and whether that need to match the system states. 

(cc: @tombentley @ShubhamRwt )